app/src/components/common/balance_token/index.tsx

```diff
@@ -0,0 +1,66 @@
+import React, { useMemo } from 'react'
+import styled from 'styled-components'
+
+import { getToken } from '../../../util/addresses'
+import { ButtonLink } from '../button_link'
+import { BigNumber } from 'ethers/utils'
+import { useAsyncDerivedValue } from '../../../hooks/useAsyncDerivedValue'
+import { formatBigNumber } from '../../../util/tools'
+import { ERC20Service } from '../../../services'
+import { useConnectedWeb3Context } from '../../../hooks/connectedWeb3'
+import { Token } from '../../../util/types'
+
+interface Props {
+  collateralId: KnownToken
+  onClickMax: (collateral: Token, collateralBalance: BigNumber) => void
+}
+
+const Wrapper = styled.span`
+  display: flex;
+  flex-direction: row;
+  width: 100%;
+  padding-bottom: 5px;
+`
+
+const BalanceTitle = styled.span`
+  margin-right: 10px;
+  margin-top: 3px;
+`
+
+export const BalanceToken = (props: Props) => {
+  const context = useConnectedWeb3Context()
+
+  const { networkId, account, library } = context
+  const { collateralId, onClickMax } = props
+
+  const collateral = getToken(networkId, collateralId)
+
+  const calculateBalanceAmount = useMemo(
+    () => async (): Promise<[BigNumber, string, string]> => {
+      const collateralService = new ERC20Service(collateral.address)
+
+      const collateralBalance = await collateralService.getCollateral(account, library)
+      return [
+        collateralBalance,
+        formatBigNumber(collateralBalance, collateral.decimals),
+        collateral.symbol,
```

Comment:
Why is the symbol and the formatted value returned by this function? They don't depend on the values that this function use, nor need to be computed asynchronously, they just pollute the function signature.

Reply:
I can code it this way (that is the logical , the right approach probably and what can you expect in this case) ...

```
export const BalanceToken = (props: Props) => {
  const context = useConnectedWeb3Context()

  const { networkId, account, library } = context
  const { collateralId, onClickMax } = props

  const collateral = getToken(networkId, collateralId)

  const calculateBalanceAmount = useMemo(
    () => async (): Promise<BigNumber> => {
      const collateralService = new ERC20Service(collateral.address)

      return collateralService.getCollateral(account, library)
    },
    [account, library, collateral],
  )

  const collateralBalanceAmountValue = useAsyncDerivedValue(
    '',
    new BigNumber(0),
    calculateBalanceAmount,
  )

  return (
    <Wrapper>
      <BalanceTitle>
        Balance {formatBigNumber(collateralBalanceAmountValue, collateral.decimals)} {collateral.symbol}.
      </BalanceTitle>
      <ButtonLink onClick={() => onClickMax(collateral, collateralBalanceAmountValue)}>Max</ButtonLink>
    </Wrapper>
  )
}
```

But if I do in this way, this is how will work:
https://www.loom.com/share/54775208575d4bf4a2935bf95518665b
You can see how the symbol and the decimals used is not in sync with calculated value.

The approach used right now work like this...
https://www.loom.com/share/13f1ad028099436d8f37baff283ea34f

I know that they don't depend on the values that this function use, but in this case it solves an small UI problem that is annoying

Reply:
Oh, I see :+1: